File: src/main/java/main/ui/renderer/ChipFreqListCellRenderer.java
package main.ui.renderer;

import main.App;
import main.puzzle.assembly.ChipFreq;
import main.ui.resource.AppImage;
import main.util.Fn;
import main.util.Ref;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import javax.swing.*;
import java.awt.*;

/**
 *
 * @author Bunnyspa
 */
public class ChipFreqListCellRenderer extends DefaultListCellRenderer {

    private final App app;
    private final Ref<Boolean> blink;

    public ChipFreqListCellRenderer(App app, Ref<Boolean> blink) {
        super();
        this.app = app;
        this.blink = blink;
    }

    @NotNull
    @Override
    public Component getListCellRendererComponent(JList<?> list, @Nullable Object value, int index, boolean isSelected, boolean cellHasFocus) {
        DefaultListCellRenderer cr = (DefaultListCellRenderer) super.getListCellRendererComponent(list, value, index, isSelected, cellHasFocus);

        cr.setHorizontalAlignment(CENTER);
        cr.setText("");

        ChipFreq cf = (ChipFreq) value;
        if (value != null) {
            cr.setIcon(AppImage.Chip.get(app, cf.chip));
            cr.setBackground(isSelected && blink.v ? Color.LIGHT_GRAY : Fn.percColor(app.orange(), app.green(), app.blue(), cf.freq, 0.0, 1.0));
        } else {
            cr.setIcon(null);
            cr.setBackground(Color.WHITE);
        }
        return cr;
    }
}

File: src/main/java/main/ui/renderer/ChipFreqListCellRenderer.kt
package main.ui.renderer

import main.App
import main.puzzle.assembly.ChipFreq
import main.ui.resource.AppImage
import main.util.Fn
import main.util.Ref
import java.awt.Color
import java.awt.Component
import javax.swing.DefaultListCellRenderer
import javax.swing.JList

/**
 *
 * @author Bunnyspa
 */
class ChipFreqListCellRenderer(private val app: App, private val blink: Ref<Boolean>) : DefaultListCellRenderer() {
    override fun getListCellRendererComponent(
        list: JList<*>?,
        value: Any?,
        index: Int,
        isSelected: Boolean,
        cellHasFocus: Boolean
    ): Component {
        val cr =
            super.getListCellRendererComponent(list, value, index, isSelected, cellHasFocus) as DefaultListCellRenderer
        cr.horizontalAlignment = CENTER
        cr.text = ""
        val cf = value as ChipFreq?
        if (value != null) {
            cr.icon = AppImage.Chip[app, cf!!.chip!!]
            cr.background = if (isSelected && blink.v) Color.LIGHT_GRAY else Fn.percColor(
                app.orange()!!,
                app.green()!!,
                app.blue()!!,
                cf.freq,
                0.0,
                1.0
            )
        } else {
            cr.icon = null
            cr.background = Color.WHITE
        }
        return cr
    }
}